src/python/module/smartredis/dataset_utils.py

```diff
@@ -0,0 +1,170 @@
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import xarray as xr
+from .dataset import Dataset
+from .util import Dtypes, exception_handler, typecheck
+from itertools import permutations
+from .error import *
+
+#----helper function -----
+def get_data(dataset,name,type):
+    return dataset.get_meta_strings(f"_xarray_{name}_{type}_names")[0].split(",")
+
+def typechecking(names,strings_name,string_name):
+    typecheck(names, strings_name, list)
+    for name in names:
+        typecheck(name, string_name, str)  
+        # Check if empty 
+        if string_name == "":
+            raise RedisRuntimeError
+        
+class DatasetConverter:  
+
+    @staticmethod
+    def add_metadata_for_xarray(dataset, data_names, dim_names, coord_names=None, attr_names=None):
+        """Extract metadata from a SmartRedis dataset and add it to 
+            dataformat specific fieldnames 
+
+            :param dataset: a Dataset instance
+            :type dataset: Dataset
+            :param data_names: variable data field name
+            :type data_names: list[str] or str 
+            :param dim_names: dimension field names.
+            :type dim_names: list[str]
+            :param coord_names: coordinate field names. Defaults to None.
+            :type coord_names: list[str], optional
+            :param attr_names: attribute field names. Defaults to None.
+            :type attr_names: list[str], optional
+        """        
+    
+        if type(data_names) == str:
+            data_names = [data_names]
+        if type(dim_names) == str:
+            dim_names = [dim_names]
+        if type(coord_names) == str:
+            coord_names = [coord_names]
+        if type(attr_names) == str:
+            attr_names = [attr_names]
+        
+        typecheck(dataset, "dataset", Dataset)   
+        typechecking(data_names, "data_names","data_name")
```

Comment:
Global: put a space after a comma.  There are about a dozen places in this file where it's missing